File: benchmarks/text/AsciiToUnicodeConversionTinyLiteral.c
#include <stdint.h>

int main(void) {
    static const unsigned char source[] = "OK";
    int32_t unicode[8] = {0};
    int64_t checksum = 0;

    for (int32_t i = 0; i < 500000; i += 1) {
        int ascii_only = 1;
        int32_t length = 0;
        for (; source[length] != '\0'; length += 1) {
            unsigned char unit = source[length];
            if ((unit & 0x80u) != 0) {
                ascii_only = 0;
                break;
            }

            unicode[length] = (int32_t)unit;
        }

        if (!ascii_only) {
            return 1;
        }

        checksum += length;
        checksum += unicode[0];
        checksum += unicode[length - 1];
    }

    return checksum == 78000000 ? 0 : 2;
}

File: benchmarks/text/AsciiToUnicodeConversionTinyLiteral.rs
fn main() {
    let source = b"OK";
    let mut unicode = [0_i32; 8];
    let mut checksum = 0_i64;

    for _ in 0_i32..500_000 {
        for (index, unit) in source.iter().copied().enumerate() {
            if (unit & 0x80) != 0 {
                std::process::exit(1);
            }

            unicode[index] = i32::from(unit);
        }

        checksum += source.len() as i64;
        checksum += i64::from(unicode[0]);
        checksum += i64::from(unicode[source.len() - 1]);
    }

    if checksum != 78_000_000 {
        std::process::exit(2);
    }
}
